File: mobile/src/types/trip.ts
import type { TripExpense } from "./expense";

export const TRIP_STATUSES = [
  "draft",
  "pending_confirmation",
  "active",
  "completed",
  "cancelled",
] as const;

export type TripStatus = (typeof TRIP_STATUSES)[number];

export interface Trip {
  _id: string;
  userId: string;
  truckRegistration: string;
  origin: string;
  destination: string;
  cargoType?: string;
  cargoWeightTons?: number;
  freightAmount: number;
  advancePaid: number;
  balanceDue: number;
  driverName?: string;
  driverPhone?: string;
  partyName?: string;
  status: TripStatus;
  expenses: TripExpense[];
  startedAt: string;
  completedAt: string | null;
  cancelledAt: string | null;
  notes?: string;
  createdAt: string;
  updatedAt: string;
  totalExpenses: number;
  netProfit: number;
}

export interface CreateTripInput {
  origin: string;
  destination: string;
  cargoType?: string;
  cargoWeightTons?: number;
  freightAmount?: number;
  advancePaid?: number;
}

File: android/app/src/main/java/com/maalsaathi/app/data/models/Trip.kt
package com.maalsaathi.app.data.models

data class Trip(
    val id: String,
    val userId: String = "",
    val organizationId: String? = null,
    val truckId: String = "",
    val driverId: String = "",
    val driverName: String = "",
    val origin: String,
    val destination: String,
    val cargoType: String = "",
    val cargoWeightTons: Double = 0.0,
    val freightAmount: Long = 0,
    val advanceAmount: Long = 0,
    val status: TripStatus,
    val startTime: Long,
    val endTime: Long? = null,
    val scheduledDate: Long? = null,
    val partyName: String = "",
    val entries: List<TripEntry> = emptyList(),
) {
    val totalExpenses: Long get() = entries.filter { it.type == EntryType.EXPENSE }.sumOf { it.amount }
    val totalIncome: Long get() = freightAmount + entries.filter { it.type == EntryType.INCOME }.sumOf { it.amount }
    val netProfit: Long get() = totalIncome - totalExpenses
    val durationMillis: Long? get() = endTime?.let { it - startTime }
}
